astropy/io/misc/pyarrow/csv.py

```diff
@@ -0,0 +1,402 @@
+    ``comment`` string is encoded to bytes using the provided encoding for comparison
+    with the file content.
+
+    Parameters
+    ----------
+    input_file : os.PathLike, str, or BinaryIO
+        The input file path or file-like object to read from.
+    comment : str
+        The comment string that identifies lines to be removed.
+    encoding : str
+        The encoding to use for the input file.
+    header_start : int or None
+        The line index where the header starts. See ``read_csv`` for details.
+    data_start : int or None
+        The line index where the data starts. See ``read_csv`` for details.
+
+    Returns
+    -------
+    output_file : io.BytesIO | None
+        A BytesIO object containing the filtered content with comment lines removed, or
+        `None` if all comment lines are at the beginning of the file.
+    header_start : int | None
+        The line index after the last comment line, or `None` if not all comment lines
+        are at the beginning of the file.
+    """
+    comment_encode = comment.encode(encoding)
+    with ExitStack() as stack:
+        if isinstance(input_file, (str, os.PathLike)):
+            input_file = stack.enter_context(open(input_file, "rb"))
+
+        if header_start is None and data_start is None:
+            idx_last_comment = -1
+            for idx, line in enumerate(input_file):
+                if line.lstrip().startswith(comment_encode):
+                    if idx - idx_last_comment == 1:
+                        idx_last_comment = idx
+                    else:
+                        break
+            else:
+                return None, idx_last_comment + 1
+
+        # If we get here, we need to read the whole file and remove comment lines and
+        # write to an output BytesIO file.
+        output = io.BytesIO()
+        for idx, line in enumerate(input_file):
+            if not line.lstrip().startswith(comment_encode):
+                output.write(line)
+
+        output.seek(0)
+
+    return output, None
+
+
+def get_convert_options(
+    include_names: list[str] | None,
+    dtypes: dict[str, "npt.DTypeLike"] | None,
+    null_values: list[str] | None,
+) -> "pyarrow.csv.ConvertOptions":
+    """
+    Generate PyArrow CSV convert options.
+
+    Parameters
+    ----------
+    include_names : list or None
+        List of column names to include in the conversion. If None, all columns are
+        included.
+    dtypes : dict or None
+        Dictionary mapping column names to their respective data types. If None, default
+        data types are used.
+    null_values : list or None, optional (default None)
+        List of strings to interpret as null values. By default, only empty strings are
+        considered as null values.
+
+    Returns
+    -------
+    pyarrow.csv.ConvertOptions
+        PyArrow CSV ConvertOptions object configured with the specified column names and
+        data types.
+    """
+    pa, csv = get_pyarrow_csv()
```

Comment:
This is the only place `pa` is used as well as `csv`. I would suggest to have `get_pyarray_csv` only return `csv`, and here add `from pyarrow import from_numpy_dtype` and use that below.

Reply:
This function is very much the same pattern as https://github.com/taldcroft/astropy/blob/061161091f861a64010f6ac9087a2fec0daa03a9/astropy/io/misc/parquet.py#L507.

My worry about changing this now is that there may be code in the future that only needs `pa`, and then you need to revert this or duplicate the import exception handling in a new function. For instance I definitely want to write a `pyarrow.ecsv` reader, and it is certainly possible that will have a function that requires a `pyarrow` import without `csv`.

Reply:
Actually, I have a different idea. Stand by... :smile:

Reply:
Fixed, hopefully.